pep-0696.rst

```diff
@@ -0,0 +1,396 @@
+
+Constraints
+~~~~~~~~~~~
+
+For constrained ``TypeVar``\ s, the default needs to be one of the
+constraints. It would be an error even if it is a subtype of one of the
+constraints.
+
+.. code:: py
+
+   TypeVar("Ok", float, str, default=float)     # Valid
+   TypeVar("Invalid", float, str, default=int)  # Invalid: expected one of float or str got int
+
+Function Defaults
+~~~~~~~~~~~~~~~~~
+
+The ``TypeVar``\ 's default should also be compatible with the
+parameter's runtime default if present. But they are erased to
+not have defaults when called.
+
+Implementation
+--------------
+
+At runtime, this would involve the following changes to
+``typing.TypeVar``:
+
+-  the type passed to default would be available as a ``__default__``
+   attribute.
+
+The following changes would be required to both ``GenericAlias``\ es:
+
+-  logic to determine the defaults required for a subscription.
+
+   -  potentially a way construct ``types.GenericAliases`` using a
+      classmethod to allow for defaults in
+      ``__class_getitem__ = classmethod(GenericAlias)``
+      i.e. ``GenericAlias.with_type_var_likes()``.
+
+      .. code:: py
+
+         # _collections_abc.py
+
+         _sentinel = object()
+
+         # NOTE: this is not actually typing.TypeVar, that's in typing.py,
+         #       this is just to trick is_typevar() in genericaliasobject.c
+         class TypeVar:
+             __module__ = "typing"
+
+             def __init__(self, name, *, default=_sentinel):
+                 self.__name__ = name
+                 self.__default__ = default
+
+         YieldT = TypeVar("YieldT")
+         SendT = TypeVar("SendT", default=None)
+         ReturnT = TypeVar("ReturnT", default=None)
+
+         class Generator(Iterable):
+             __class_getitem__ = GenericAlias.with_type_var_likes(YieldT, SendT, ReturnT)
+
+-  ideally, logic to determine if subscription (like
+   ``Generic[T, DefaultT]``) would be valid.
+
+A reference implementation of the type checker can be found at
+https://github.com/Gobot1234/mypy/tree/TypeVar-defaults
+
+Interaction with PEP 695
+------------------------
+
+If this PEP were to be accepted, amendments to :pep:`695` could be made to
+allow for specifying defaults for type parameters using the new syntax.
+Specifying a default should be done using the "=" operator inside of the
+square brackets like so:
+
+.. code:: py
+
+   class Foo[T = str]: ...
+
+   def bar[U = int](): ...
+
+This functionality was included in the initial draft of PEP 695 but was
+removed due to scope creep.
+
+Grammar Changes
+~~~~~~~~~~~~~~~
+
+::
+
+      type_param:
+         | a=NAME b=[type_param_bound] d=[type_param_default]
+         | a=NAME c=[type_param_constraint] d=[type_param_default]
+         | '*' a=NAME d=[type_param_default]
+         | '**' a=NAME d=[type_param_default]
+
+      type_param_default: '=' e=expression
+
+This would mean that ``TypeVarLikes`` with defaults proceeding those
```

Comment:
I think we should decide now whether to include default support for TypeVarTuple and ParamSpec and be consistent in our recommendation. Right now, this section implies that they will be supported, but they are not specified in the rest of this PEP.

I'm of two minds here. I don't see a strong need for defaults for TypeVarTuple and ParamSpec. You included this in an earlier draft of this PEP, and I gave you feedback at the time that we might want to remove it to reduce complexity. However, if we're going to add dedicated syntax for defaults (the `=` extension to PEP 695), it's probably better to add the support now rather than piecemeal over time.

Reply:
@JelleZijlstra any thoughts on this point? I'd personally like to re-add the support for TypeVarTuple and ParamSpec cause I don't see any reason not to if PEP 695 gets accepted and even if it gets rejected I'd draft a specification up for it at some point anyway.

Reply:
I'm not opposed, but we'll have to specify clearly what kinds of defaults are allowed. For ParamSpec, it's not obvious to me how defaults should work at all, because the solution for a ParamSpec isn't directly expressible as a type. For TypeVarTuple, I suppose we'd require the default to be a tuple type (e.g. `tuple[int, str]` or `tuple[float, ...]`).

We could also allow defaults for these in the runtime syntax, but leave the semantics unspecified.

Reply:
I think the allowed defaults for ParamSpec and TypeVarTuple fall out naturally from the expression forms that are supported when providing type arguments.

For ParamSpec, that means ellipsis or a list of types (e.g. `[int, str, int]`). If no default is specified, the default is effectively `...` (as it is today).

For TypeVarTuple, it would be an unpacked tuple expression (e.g. `*tuple[int, str]`, `*tuple[int, ...]`, `*tuple[int, *tuple[str, ...], int]`, or `Unpack[tuple[int]]`). If no default is specified, the default is effectively `*tuple[Any, ...]` (as it is today).